gusto/time_discretisation/sdc.py

```diff
@@ -0,0 +1,535 @@
+y_m = y_n + sum[j=1,M] q_mj*F(y_j)
+where q_mj can be found by integrating Lagrange polynomials. This is similar to
+how Runge-Kutta methods are formed.
+
+In matrix form this equation is:
+(I - dt*Q*F)(y)=y_n
+
+Computing y by Picard iteration through k we get:
+y^(k+1)=y^k + (y_n - (I - dt*Q*F)(y^k))
+
+Finally, to get our SDC method we precondition this system, using some approximation
+of Q Q_delta:
+(I - dt*Q_delta*F)(y^(k+1)) = y_n + dt*(Q - Q_delta)F(y^k)
+
+The zero-to-node (Z2N) formulation is then:
+y_m^(k+1) = y_n + sum(j=1,M) q'_mj*(F(y_j^(k+1)) - F(y_j^k))
+            + sum(j=1,M) q_mj*F(y_(m-1)^k)
+for entires q_mj in Q and q'_mj in Q_delta.
+
+Node-wise from previous quadrature node (N2N formulation), the implicit SDC calculation is:
+y_m^(k+1) = y_(m-1)^(k+1) + dtau_m*(F(y_(m)^(k+1)) - F(y_(m)^k))
+            + sum(j=1,M) s_mj*F(y_(m-1)^k)
+where s_mj = q_mj - q_(m-1)j for entires q_ik in Q.
+
+
+Key choices in our SDC method are:
+- Choice of quadrature node type (e.g. gauss-lobatto)
+- Number of quadrature nodes
+- Number of iterations - each iteration increases the order of accuracy up to
+  the order of the underlying quadrature
+- Choice of Q_delta (e.g. Forward Euler, Backward Euler, LU-trick)
+- How to get initial solution on quadrature nodes
+"""
+
+from abc import ABCMeta
+import numpy as np
+from firedrake import (
+    Function, NonlinearVariationalProblem, TestFunction, TestFunctions,
+    NonlinearVariationalSolver, Constant
+)
+from firedrake.fml import (
+    replace_subject, replace_test_function, all_terms, drop
+)
+from firedrake.utils import cached_property
+from gusto.time_discretisation.wrappers import *
+from gusto.time_discretisation.time_discretisation import wrapper_apply
+from gusto.core.labels import (time_derivative, implicit, explicit)
+
+from qmat import genQCoeffs, genQDeltaCoeffs
+
+__all__ = ["SDC"]
+
+
+class SDC(object, metaclass=ABCMeta):
+    """Class for Spectral Deferred Correction schemes."""
+
+    def __init__(self, base_scheme, domain, M, maxk, quad_type, node_type, qdelta_imp, qdelta_exp,
+                 formulation="N2N", field_name=None,
+                 linear_solver_parameters=None, nonlinear_solver_parameters=None, final_update=True,
+                 limiter=None, options=None, initial_guess="base"):
+        """
+        Initialise SDC object
+        Args:
+            base_scheme (:class:`TimeDiscretisation`): Base time stepping scheme to get first guess of solution on
+                quadrature nodes.
+            domain (:class:`Domain`): the model's domain object, containing the
+                mesh and the compatible function spaces.
+            M (int): Number of quadrature nodes to compute spectral integration over
+            maxk (int): Max number of correction interations
+            quad_type (str): Type of quadrature to be used. Options are
+            GAUSS, RADAU-LEFT, RADAU-RIGHT and LOBATTO
```

Comment:
Could you indent these overflowing comments?

Reply:
Done!